runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+			panic("unimplemented")
+		default:
+			return fmt.Errorf("unknown backup format %q", meta.Format)
+		}
+	}
+	// create views after attaching all the DBs since views can depend on other tables
+	for _, view := range views {
+		_, err := db.ExecContext(ctx, view)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *db) readTableVersion(name string) (string, bool, error) {
+	return tableVersion(d.readPath, name)
+}
+
+func (d *db) writeTableVersion(name string) (string, bool, error) {
+	return tableVersion(d.writePath, name)
+}
+
+func execIncrementalInsert(ctx context.Context, conn *sqlx.Conn, safeTableName, query string, opts *InsertTableOptions) error {
+	var byNameClause string
+	if opts.ByName {
+		byNameClause = "BY NAME"
+	}
+
+	if opts.Strategy == IncrementalStrategyAppend {
+		_, err := conn.ExecContext(ctx, fmt.Sprintf("INSERT INTO %s %s (%s\n)", safeTableName, byNameClause, query))
+		return err
+	}
+
+	if opts.Strategy == IncrementalStrategyMerge {
+		// Create a temporary table with the new data
+		tmp := uuid.New().String()
+		_, err := conn.ExecContext(ctx, fmt.Sprintf("CREATE TEMPORARY TABLE %s AS (%s\n)", safeSQLName(tmp), query))
+		if err != nil {
+			return err
+		}
+
+		// check the count of the new data
+		// skip if the count is 0
+		// if there was no data in the empty file then the detected schema can be different from the current schema which leads to errors or performance issues
+		res := conn.QueryRowContext(ctx, fmt.Sprintf("SELECT COUNT(*) == 0 FROM %s", safeSQLName(tmp)))
+		var empty bool
+		if err := res.Scan(&empty); err != nil {
+			return err
+		}
+
+		if empty {
+			return nil
+		}
+
+		// Drop the rows from the target table where the unique key is present in the temporary table
+		where := ""
+		for i, key := range opts.UniqueKey {
+			key = safeSQLName(key)
+			if i != 0 {
+				where += " AND "
+			}
+			where += fmt.Sprintf("base.%s IS NOT DISTINCT FROM tmp.%s", key, key)
+		}
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("DELETE FROM %s base WHERE EXISTS (SELECT 1 FROM %s tmp WHERE %s)", safeTableName, safeSQLName(tmp), where))
+		if err != nil {
+			return err
+		}
+
+		// Insert the new data into the target table
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("INSERT INTO %s %s SELECT * FROM %s", safeTableName, byNameClause, safeSQLName(tmp)))
+		return err
+	}
+
+	return fmt.Errorf("incremental insert strategy %q not supported", opts.Strategy)
+}
+
+func tableVersion(path, name string) (string, bool, error) {
+	pathToFile := filepath.Join(path, name, "version.txt")
+	contents, err := os.ReadFile(pathToFile)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return "", false, nil
+		}
+		return "", false, err
+	}
+	return strings.TrimSpace(string(contents)), true, nil
+}
+
+func newVersion() string {
+	return strconv.FormatInt(time.Now().UnixMilli(), 10)
+}
+
+type meta struct {
+	ViewSQL string
+	Format  BackupFormat
+}
+
+func writeMeta(path string, meta meta) error {
```

Comment:
1. Is this for each table? If there's a meta file for each table, then why is a separate `version.txt` needed – we should be economical with the number of files where possible, object stores bill per operation (and it increases the chance of inconsistency).
2. By the way, have you considered having a single manifest/meta file for the entire DB instead? It would reduce the number of files to check to know if you are in sync or not, and might make it easier to resolve conflicts in case of concurrent writes (which is going to be a risk).

Reply:
1. This is based on our discussed directory structure. The meta is per version. It has information about the view SQL and backup format. `version.txt` specifies which version to use. `Meta` being per version can handle cases when a table is recreated as a view. Given we don't have the backup format as of now we can avoid creating it for tables and create it just for views which can reduce the number of files.
2. I actually thought about keeping all versions in a single JSON file but then I remembered we had some discussions about this when implementing `external table storage` and we agreed on keeping it specific to a table(I think it was for easy to visualize versions but I am not sure).

Reply:
> Meta being per version can handle cases when a table is recreated as a view

I think this could also be solved with a single meta file per table, right? You just update the per-table file with the view and its version number. Like this:
```bash
# v1 is a table:
v1.db
meta.json # {version: 1}

# v2 is a view:
meta.json # {version: 2, view_sql: "SELECT ..."}
```

> and create it just for views which can reduce the number of files

If there are other reasons for keeping per-version meta files, then I think this at least sounds better

> I actually thought about keeping all versions in a single JSON file but then I remembered we had some discussions about this when implementing external table storage and we agreed on keeping it specific to a table

Yeah I think that made sense for local external table storage because a) file operations are more lightweight, b) you never have to think about consistency in the face of concurrent readers/writers.

When backed by a bucket, if possible, it's nice to reduce the number of files that need to be listed/updated/downloaded. I think it also makes various consistency issues easier to handle – obviously Iceberg is a different beast with a much greater scope, but take a look at how they use manifest files to enable consistent reads/writes on an object store.

Reply:
Thanks for pointing to the `Iceberg` reference. 
So I am now creating a `meta.json` file now for each table with version and some other metadata. This will also help when concurrent ETL writers update tables

I think we can also create a single metadata file for entire DB for quick checks as you mentioned. But I think this can be taken up in future when we separate reads and writes. As of now `writeDirty` bool in db will make sure we never read the metadata files from object store except during boot up or edge cases. 
I also think each write updating a single file may lead to contention and too many failures (due to ETag check failures). Need to dig deeper on how Iceberg manages this.

Reply:
> I also think each write updating a single file may lead to contention and too many failures (due to ETag check failures). Need to dig deeper on how Iceberg manages this.

They basically just retry (and design in such a way that contention doesn't happen too much and is mostly lightweight to retry): https://iceberg.apache.org/docs/1.6.0/reliability/#concurrent-write-operations